### ISSUE TITLE
WAMP-CRA is potentially unsafe

### ISSUE THREAD

**Issue by Paranaix** [tavendo/WAMP#128]
![](http://puu.sh/f6fX9/e3910a6519.png)

**Congratulations, you just got admin access to my service ;)**

To make it simply: Because the client has to sign a challenge on clientside when using WAMP-CRA the server is unable to hash the user supplied value again on serverside, giving every attacker which knows the stored hash access. Simply sign the challenge directly with the stolen hash -> "Access granted".

**Can Three-legged authentication fix this issue?** 
No.
In the case we send a (prehashed) secret and the challenge to on of our servers again, where yonder server will hash the secret with the stored salt and sign the challenge, we have the problem, that we loose all the advantages of WAMP-CRA as a attacker might be able to get a glimpse on the prehashed secret. Therefor this method would basically to the classic approach of directly sending a prehashed secret to a server and then hash it with the salt on serverside. In that case we would not need WAMP-CRA and could directly go with ticket based authentication.

**Proposed solution**
Authentication in 3 instead of 2 steps. Of course there might many different possible solutions, but I atleast want to show ONE of them.

Let **s** be a secret only known to the user and **H1** and **H2** two hash functions with **H2(s) = H1( f(s) )** where **f** is a function which deforms the originally password in some unspecified way so that **H1(s) != H2(s)** is true.

The authentication process starts regulary with the client announcing its possible authmethods and the server choosing one. In the case the server chooses this proposed method (we could maybe call it *Advanced WAMP-CRA*) he sends a challenge and the client simply responses with **H1(s)**. We assume that an attacker might be able to intercept this hash value, This security hole is fixed with the next authentication step. However it is apparent, that **H1** has to be a strong enough cryptographic hash function so that the attacker can't get **s**. If the server now receives **H1(s)** from the client he will hash ist again with the salt and compare it to a hash **a = H1( H1(s) + salt )** e.g stored in a database effectively fixing the problem described in this issue. If both hashes matchs we will send another challenge to the client this time containing the JSON challenge string also used in regular WAMP-CRA authentication. The client will then sign this challenge with **H2(H1(s) + salt)** and the server will likewise compare the sinature against a stored value **b = H2(H1(s) + salt** (notice: **H1(s)** should just acts as a prehashed password here because we shouldnt ever send plain password to any server, no matter how trustworthy). If both signatures match, authentication is complete.

**Why if f(s) needed**
The purpose of **f** is to being abe to create 2 distinct hash values from the same hash function. By using the same underlying hash function we prevent any possible collisions (assuming that a secure hash function is used). A simple implementation of **f** could e.g just append some characters (basically a salt function) or switch cases or just replace every character with the next one in the alphabet. The only assertion to be true is **f(s) != s**

**Is this secure?**
This method is basically a combination of both authentication methods and thus fixes the issues of both. The only way for an attacker to get access now is to 1) intercept **H1(s)** 2) also steal **b = H2(H1(s) + salt)**. An attacker which only managed to achieve one is unable to gain access.

**Comment by oberstet** [comment]
@Paranaix Not sure if I get you here: are you saying that with salted WAMP-CRA, the _sending of the salt over the wire_ allows a man-in-the-middle to create a valid signature?

**Comment by Paranaix** [comment]
The problem is that the actual hashing takes place on the client (due to the nature of WAMP-CRA) and not on the server. Therefor we can't really differentiate whether the peer knows the actual secret (e.g a password even we don't know) or only the hash. If an attacker would get access to a db he could authenticate as any user he wants without even cracking a single hash. I found this by accident as I tested my authentication and accidantally signed the challenge directly with the hash from the db.

Im not saying that WAMP-CRA is unsecure in general, If I had to choose between sending a plain password over the wire or having trouble if someone stole data from my db I would definitely go with later. However I believe it might be possible to fix both issues by combining both appoaches as I already stated. The proposed solution is however only that: proposed. The main concern of this issue is to make you aware of a potential security hole.

**Comment by oberstet** [comment]
As this will likely be a longer thread, here is the short takeaway first:

WAMP-CRA does provide:

1. protection against man-in-the-middle (even with no transport encryption)
2. protection against replay-attacks (and request delaying)
3. does NOT protect against brute-force, dictionary attacks
4. does NOT protect against password collisions
5. does NOT protect against credential database theft

WAMP-CRA with salting _additionally_ provides:

1. protection against brute-force, dictionary attacks
2. protection against password collisions

**However, yes, WAMP-CRA with salting does still NOT protect against credential database theft.**

====

**Discussion**

Yes, with WAMP-CRA salted, the key derived from the original password (and the salt) via PBKDF2 essentially acts as a new, shared secret.

It is still a shared _secret_ system. And if the shared secret (the key derived of the original password) is disclosed to a 3rd party (database dump), you a screwed.

However, it is still better than storing passwords plainly, since it prohibits brute-force dictionary attacks (password crackers).

How many users will use a weak, commonly used "passwords"? How many users will end up using the *same* password? With salted WAMP-CRA, all derived keys are different (since of the salting).

It will not protect you against an attacker that is able to dump your database. It will not protect the user from an attacker that is able to read any browser/client side cache / local storage containing the secret.

I think this deserves some context to put into perspective, and I think we can improve even further.

WAMP-CRA with no salting at all is already more secure than both http://en.wikipedia.org/wiki/Basic_access_authentication and http://en.wikipedia.org/wiki/Digest_access_authentication.

HTTP Basic Auth sends cleartext passwords over the wire. Well, bad. HTTP Digest Auth is a challenge-response scheme roughly similar to WAMP-CRA, but it uses MD5, not HMAC-SHA256 as WAMP-CRA. It also includes a nonce, but no timestamp.

It is also more secure than the 99% login scheme used on the Web: unsalted passwords directly sent over a transport level encrypted connection and cookies.

Of course you can run WAMP-CRA over TLS also. And with Crossbar.io, you can also use Cookie based authentication for directly authenticating a previously authenticated user.

If someone breaks TLS, with WAMP-CRA, you still have protection. You are screwed using Cookies then with both (cookie theft in general).

**Comment by oberstet** [comment]
> Let s be a secret only known to the user and H1 and H2 two hash functions with H2(s) = H1( f(s) ) where f is a function which deforms the originally password in some unspecified way so that H1(s) != H2(s) is true.

There are problems with this proposal:

1. It is unclear if H2 and H1 with those properties exist, and we need concrete H1 and H2 that can be implemented
2. Even given 1., the formal security of the scheme is unclear
3. It relies on ad-hoc, invented crypto primitives

Rather than inventing something new (which we deliberately avoid with WAMP-CRA - which uses well know crypto primitives in standard combination), I think we should follow recent developments from the crypto community to improve further on "WAMP-CRA salted".

There are 2 schemes I am aware of that might be suitable:

1. **SRP** http://en.wikipedia.org/wiki/Secure_Remote_Password_protocol
2. **SCRAM** http://en.wikipedia.org/wiki/Salted_Challenge_Response_Authentication_Mechanism

From looking at SCRAM, this claims to provide protection against these attack vectors:

1. Eavesdropping
2. Replay
3. Database Compromise

and also against dictionary attacks and password collisions.

It additionally claims to protect against impersonating a server (after the original server had signed up a user).

Here are some references:

* http://en.wikipedia.org/wiki/Salted_Challenge_Response_Authentication_Mechanism
* https://tools.ietf.org/html/rfc5802
* http://www.mongodb.com/blog/post/improved-password-based-authentication-mongodb-30-scram-explained-part-1
* http://www.mongodb.com/blog/post/improved-password-based-authentication-mongodb-30-scram-explained-part-2

====

The good news is, with the WAMP messages `HELLO`, `CHALLENGE`, `AUTHENTICATE` and `WELCOME` (and the extensibility built in those messages), we have the required message flow in place without changing WAMP.

However, I don't think it makes sense to put this into WAMP-CRA as it exists.

**I think we should come up with a mapping of SCRAM to WAMP: WAMP-SCRAM (`wamp-scram`)**

This should be pretty straight-foward, and we can reuse an already designed, scrutenized scheme.

We also do not need new crypto primitives, as the following will do fine:

* SHA256
* HMAC
* PBKDF2
* XOR

**Comment by Paranaix** [comment]
I agree with you that using an already established method is way better than comming up with something on our own. *SCRAM* looks very promising. This should definitely be a new separate authmethod, It should be however made clear, that *SCRAM* is favored over *WAMP-CRA*. If a server and peer supports both, the server should always choose *SCRAM* over *WAMP-CRA*. *WAMP-CRA* would take the role of a fallback method.

However I think that your statement that `we have the required message flow in place without changing WAMP.` is wrong. As `HELLO` is used to agree on an (at that time yet unknown) authentication method, using it to transmitting data unique to the a specific authmethod feels highly wrong. The same applies to `WELCOME`: IMO `WELCOME` indicates a successful authentication, however as far as I undrstood the *SCRAM* protocol, the client still has to check the server proof.

I suggest reusing the `CHALLENGE` message and allowing authentication methods to send them back-and-forth as much as required (opposed to two times as it is now). 

Should we maybe create new issue (with a backlink to this one) for the concrete elaboration of the specification for *wamp-scram*?

**Comment by oberstet** [comment]
Yes, I think a new issue would make sense. Let's discuss any details there.

Rgd this issue here: if you agree, I'd leave this open and rename this to "Describe security aspects of WAMP-CRA" to add text to the spec that discusses what is, and what is not provided by WAMP-CRA. Since: I would say, the lack of protection against database theft is not an actual security hole, but a security limit. The former would be the case when a security aspect that WAMP-CRA announces would be broken.

**Comment by ecorm** [comment]
From a data security perspective, the WAMP-CRA scheme as it currently stands is only marginally better than storing the passwords in plain text.

Except for the Heartbleed exploit, most of the major security breaches you hear about are data breaches.

**Comment by ecorm** [comment]
I believe the WAMP-CRA vulnerability would be classified as [Pass the Hash](https://en.wikipedia.org/wiki/Pass_the_hash).

**Comment by ecorm** [comment]
@oberstet 

> Since: I would say, the lack of protection against database theft is not an actual security hole, but a security limit.

I'm afraid I disagree about it not being a security hole. The whole point of password hashing is to safeguard against database theft. Every trustworthy article I've been reading about password hashing says that is *must* be done by the server, and not the client.

- https://security.stackexchange.com/questions/53594/why-is-client-side-hashing-of-a-password-so-uncommon
- https://cybergibbons.com/security-2/stop-doing-client-side-password-hashing/
- https://security.stackexchange.com/questions/8596/https-security-should-password-be-hashed-server-side-or-client-side

If the goal of WAMP-CRA is to provide another layer of protection against a transport lacking SSL, then this should be done in a manner that doesn't bypass the server needing to compute the password hashes.

Based on my research on this subject, it appears that dynamic ticket-based authentication over SSL with the authenticator computing the hash using a modern algorithm would be more secure than WAMP-CRA.
